statefun-kafka-io/src/main/java/org/apache/flink/statefun/sdk/kafka/KafkaIngressStartupPosition.java

```diff
@@ -0,0 +1,153 @@
+  public static KafkaIngressStartupPosition fromLatest() {
+    return new LatestPosition();
+  }
+
+  /**
+   * Start consuming from a specified set of offsets.
+   *
+   * <p>If a specified offset does not exist for a partition, the position for that partition will
+   * fallback to the reset position configured via {@link
+   * KafkaIngressBuilder#withAutoResetPosition(KafkaIngressAutoResetPosition)}.
+   *
+   * @param specificOffsets map of specific set of offsets.
+   */
+  public static KafkaIngressStartupPosition fromSpecificOffsets(
+      Map<KafkaTopicPartition, Long> specificOffsets) {
+    if (specificOffsets == null || specificOffsets.isEmpty()) {
+      throw new IllegalArgumentException("Provided specific offsets must not be empty.");
+    }
+    return new SpecificOffsetsPosition(specificOffsets);
+  }
+
+  /**
+   * Start consuming from offsets with ingestion timestamps after or equal to a specified {@link
+   * Date}.
+   *
+   * <p>If a Kafka partition does not have any records with ingestion timestamps after or equal to
+   * the specified date, the position for that partition will fallback to the reset position
+   * configured via {@link
+   * KafkaIngressBuilder#withAutoResetPosition(KafkaIngressAutoResetPosition)}.
+   */
+  public static KafkaIngressStartupPosition fromDate(Date date) {
+    return new DatePosition(date);
+  }
+
+  /** Checks whether this position is configured using committed consumer group offsets in Kafka. */
+  public boolean isGroupOffsets() {
+    return getClass() == GroupOffsetsPosition.class;
+  }
+
+  /** Checks whether this position is configured using the earliest offset. */
+  public boolean isEarliest() {
+    return getClass() == EarliestPosition.class;
+  }
+
+  /** Checks whether this position is configured using the latest offset. */
+  public boolean isLatest() {
+    return getClass() == LatestPosition.class;
+  }
+
+  /** Checks whether this position is configured using specific offsets. */
+  public boolean isSpecificOffsets() {
+    return getClass() == SpecificOffsetsPosition.class;
+  }
+
+  /** Checks whether this position is configured using a date. */
+  public boolean isDate() {
+    return getClass() == DatePosition.class;
+  }
+
+  /** Returns this position as a {@link SpecificOffsetsPosition}. */
+  public SpecificOffsetsPosition asSpecificOffsets() {
+    if (!isSpecificOffsets()) {
+      throw new IllegalStateException(
+          "This is not a startup position configured using specific offsets.");
+    }
+    return (SpecificOffsetsPosition) this;
+  }
+
+  /** Returns this position as a {@link DatePosition}. */
+  public DatePosition asDate() {
+    if (!isDate()) {
+      throw new IllegalStateException("This is not a startup position configured using a Date.");
+    }
+    return (DatePosition) this;
+  }
+
+  public static class GroupOffsetsPosition extends KafkaIngressStartupPosition {}
```

Comment:
I think it would be better if all of these would be final, with a private constructor.